File: base/src/main/java/io/quarkus/code/service/PlatformService.java
package io.quarkus.code.service;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.SortedSet;
import java.util.concurrent.atomic.AtomicReference;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.regex.Pattern;

import io.quarkus.code.model.Preset;
import io.quarkus.code.model.ProjectDefinition;
import io.quarkus.code.model.Stream;
import io.quarkus.devtools.commands.data.QuarkusCommandException;
import io.quarkus.devtools.project.JavaVersion;
import io.quarkus.devtools.project.QuarkusProjectHelper;
import io.quarkus.logging.Log;
import io.quarkus.registry.Constants;
import io.quarkus.registry.ExtensionCatalogResolver;
import io.quarkus.registry.catalog.PlatformRelease;
import io.quarkus.runtime.LaunchMode;
import jakarta.enterprise.event.Observes;
import jakarta.enterprise.inject.Instance;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import io.quarkus.code.config.PlatformConfig;
import io.quarkus.code.model.CodeQuarkusExtension;
import io.quarkus.registry.RegistryResolutionException;
import io.quarkus.registry.catalog.ExtensionCatalog;
import io.quarkus.registry.catalog.Platform;
import io.quarkus.registry.catalog.PlatformCatalog;
import io.quarkus.registry.catalog.PlatformStream;
import io.quarkus.runtime.StartupEvent;
import io.quarkus.scheduler.Scheduled;

import static io.quarkus.code.misc.QuarkusExtensionUtils.processExtensions;
import static io.quarkus.devtools.project.JavaVersion.getCompatibleLTSVersions;
import static io.quarkus.platform.catalog.processor.CatalogProcessor.getMinimumJavaVersion;
import static io.quarkus.platform.catalog.processor.CatalogProcessor.getRecommendedJavaVersion;

@Singleton
public class PlatformService {

    public static final List<Preset> DEFAULT_PRESETS = List.of(
            // Some presets are duplicated to support platforms before and after the Big Reactive Renaming
            new Preset("rest-service", "REST service",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/rest.svg",
                    List.of("io.quarkus:quarkus-rest")),
            new Preset("rest-service", "REST service",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/rest.svg",
                    List.of("io.quarkus:quarkus-resteasy-reactive")),
            new Preset("db-service", "REST service with database",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/db-service.svg",
                    List.of("io.quarkus:quarkus-rest", "io.quarkus:quarkus-rest-jackson",
                            "io.quarkus:quarkus-hibernate-orm-panache", "io.quarkus:quarkus-jdbc-postgresql")),
            new Preset("db-service", "REST service with database",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/db-service.svg",
                    List.of("io.quarkus:quarkus-resteasy-reactive", "io.quarkus:quarkus-resteasy-reactive-jackson",
                            "io.quarkus:quarkus-hibernate-orm-panache", "io.quarkus:quarkus-jdbc-postgresql")),
            new Preset("event-driven-kafka", "Event driven service with Kafka",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/event-driven-kafka.svg",
                    List.of("io.quarkus:quarkus-messaging-kafka")),
            new Preset("event-driven-kafka", "Event driven service with Kafka",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/event-driven-kafka.svg",
                    List.of("io.quarkus:quarkus-smallrye-reactive-messaging-kafka")),
            new Preset("cli", "Command-line tool",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/cli.svg",
                    List.of("io.quarkus:quarkus-picocli")),
            new Preset("webapp-mvc", "Web app with Model-View-Controller",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/webapp-mvc.svg",
                    List.of("io.quarkiverse.renarde:quarkus-renarde", "io.quarkiverse.web-bundler:quarkus-web-bundler")),
            new Preset("webapp-npm", "Web app with NPM UI",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/webapp-npm.svg",
                    List.of("io.quarkus:quarkus-rest", "io.quarkus:quarkus-rest-jackson",
                            "io.quarkiverse.quinoa:quarkus-quinoa")),
            new Preset("webapp-npm", "Web app with NPM UI",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/webapp-npm.svg",
                    List.of("io.quarkus:quarkus-resteasy-reactive", "io.quarkus:quarkus-resteasy-reactive-jackson",
                            "io.quarkiverse.quinoa:quarkus-quinoa")),
            new Preset("webapp-qute", "Web app with ServerSide Rendering",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/webapp-qute.svg",
                    List.of("io.quarkiverse.qute.web:quarkus-qute-web", "io.quarkiverse.web-bundler:quarkus-web-bundler")),
            new Preset("ai-infused", "AI Infused service",
                    "https://raw.githubusercontent.com/quarkusio/code.quarkus.io/main/base/assets/icons/presets/ai-infused.svg",
                    List.of("io.quarkiverse.langchain4j:quarkus-langchain4j-openai",
                            "io.quarkiverse.langchain4j:quarkus-langchain4j-easy-rag")));
    public static final Pattern NUMERIC_PATTERN = Pattern.compile("\\d+");

    @Inject
    private PlatformConfig platformConfig;

    @Inject
    private QuarkusProjectService projectService;

    @Inject
    Instance<PlatformOverride> platformOverride;

    private final ExtensionCatalogResolver catalogResolver;
    private final AtomicReference<PlatformServiceCache> platformServiceCacheRef = new AtomicReference<>();

    public void onStart(@Observes StartupEvent e) {
        reload();
    }

    public PlatformService() throws RegistryResolutionException {
        this.catalogResolver = QuarkusProjectHelper.getCatalogResolver();
    }

    @Scheduled(cron = "{io.quarkus.code.quarkus-platforms.reload-cron-expr}")
    public void reload() {
        try {
            reloadPlatformServiceCache();
        } catch (RegistryResolutionException e) {
            Log.warnf(e, "Could not resolve catalogs [%s]", e.getLocalizedMessage());
        } catch (Exception e) {
            Log.warnf(e, "Could not reload catalogs [%s]", e.getLocalizedMessage());
        }
    }

    public boolean isLoaded() {
        return platformServiceCacheRef.get() != null && !recommendedCodeQuarkusExtensions().isEmpty();
    }

    public PlatformServiceCache platformsCache() {
        PlatformServiceCache cache = platformServiceCacheRef.get();
        if (cache == null) {
            throw new IllegalStateException("Platforms cache must not be used if not loaded");
        }
        return cache;
    }

    public List<Preset> presets() {
        return platformOverride.isResolvable() ? platformOverride.get().presets() : DEFAULT_PRESETS;
    }

    public PlatformInfo recommendedPlatformInfo() {
        return platformInfo(null);
    }

    public LocalDateTime cacheLastUpdated() {
        return platformsCache().cacheLastUpdated();
    }

    public PlatformCatalog platformCatalog() {
        return platformsCache().platformCatalog();
    }

    public List<CodeQuarkusExtension> recommendedCodeQuarkusExtensions() {
        return codeQuarkusExtensions(recommendedStreamKey());
    }

    public String recommendedStreamKey() {
        return platformsCache().recommendedStreamKey();
    }

    public List<Stream> streams() {
        return platformServiceCacheRef.get().streams();
    }

    public Set<String> streamKeys() {
        return platformServiceCacheRef.get().streamCatalogMap().keySet();
    }

    public List<CodeQuarkusExtension> codeQuarkusExtensions(String platformKey, String streamId) {
        String key = createStreamKey(platformKey, streamId);
        return codeQuarkusExtensions(key);
    }

    public List<CodeQuarkusExtension> codeQuarkusExtensions(String streamKey) {
        return platformInfo(streamKey).codeQuarkusExtensions();
    }

    public PlatformInfo platformInfo(String platformKey, String streamId) {
        String key = createStreamKey(platformKey, streamId);
        return platformInfo(key);
    }

    public PlatformInfo platformInfo(String streamKey) {
        String normalizedStreamKey = normalizeStreamKey(streamKey);
        Map<String, PlatformInfo> streamCatalogMap = platformServiceCacheRef.get().streamCatalogMap();
        if (streamCatalogMap.containsKey(normalizedStreamKey)) {
            return streamCatalogMap.get(normalizedStreamKey);
        } else {
            throw new IllegalArgumentException("Invalid streamKey: " + streamKey);
        }
    }

    private PlatformOverride getPlatformOverride() {
        return platformOverride.isResolvable() ? platformOverride.get() : PlatformOverride.DEFAULT_PLATFORM_OVERRIDE;
    }

    private void reloadPlatformServiceCache() throws RegistryResolutionException, IOException, QuarkusCommandException {
        catalogResolver.clearRegistryCache();
        PlatformCatalog platformCatalog;
        if (platformConfig.getRegistryId().isEmpty()) {
            platformCatalog = catalogResolver.resolvePlatformCatalog();
        } else {
            platformCatalog = catalogResolver.resolvePlatformCatalogFromRegistry(platformConfig.getRegistryId().get());
        }
        Map<String, PlatformInfo> updatedStreamCatalogMap = new HashMap<>();
        if (platformCatalog == null || platformCatalog.getMetadata() == null
                || platformCatalog.getPlatforms() == null) {
            throw new RuntimeException("Platform catalog not found");
        }

        String platformTimestamp = platformCatalog.getMetadata().get(Constants.LAST_UPDATED).toString();
        if (platformTimestamp.isBlank()) {
            throw new RuntimeException("Platform last updated date is empty");
        }
        if (platformServiceCacheRef.get() != null
                && platformServiceCacheRef.get().platformTimestamp().equals(platformTimestamp)) {
            LOG.log(Level.INFO, "The platform cache is up to date with the registry");
            return;
        }
        Collection<Platform> platforms = platformCatalog.getPlatforms();
        List<Stream> streams = new ArrayList<>();
        for (Platform platform : platforms) {
            for (PlatformStream stream : platform.getStreams()) {
                PlatformRelease recommendedRelease = stream.getRecommendedRelease();
                ExtensionCatalog extensionCatalog = catalogResolver
                        .resolveExtensionCatalog(recommendedRelease.getMemberBoms());
                List<CodeQuarkusExtension> codeQuarkusExtensions = processExtensions(extensionCatalog,
                        getPlatformOverride());
                String platformKey = platform.getPlatformKey();
                String streamId = stream.getId();
                String streamKey = createStreamKey(platformKey, streamId);
                boolean lts = (boolean) stream.getMetadata().get("lts");
                String minimumJavaVersion = getMinimumJavaVersion(extensionCatalog);

                SortedSet<Integer> compatibleJavaLTSVersions = getCompatibleLTSVersions(
                        new JavaVersion(minimumJavaVersion));
                int recommendedJavaVersion = Optional.ofNullable(getRecommendedJavaVersion(extensionCatalog))
                        .map(Integer::parseInt).orElse(compatibleJavaLTSVersions.stream().findFirst().orElseThrow());
                String quarkusCoreVersion = stream.getRecommendedRelease().getQuarkusCoreVersion();
                boolean recommended = stream.getId().equals(platform.getRecommendedStream().getId());
                final String platformVersion = stream.getRecommendedRelease().getVersion().toString();
                Stream streamInfo = Stream.builder()
                        .key(streamKey)
                        .quarkusCoreVersion(quarkusCoreVersion)
                        .javaCompatibility(
                                new Stream.JavaCompatibility(compatibleJavaLTSVersions, recommendedJavaVersion))
                        .lts(lts)
                        .platformVersion(platformVersion)
                        .recommended(recommended)
                        .status(getStreamStatus(quarkusCoreVersion))
                        .build();
                PlatformInfo platformInfo = new PlatformInfo(
                        platformKey,
                        streamInfo,
                        quarkusCoreVersion,
                        platformVersion,
                        recommended,
                        codeQuarkusExtensions,
                        extensionCatalog);
                if (!codeQuarkusExtensions.isEmpty()) {
                    streams.add(streamInfo);
                    updatedStreamCatalogMap.put(streamKey, platformInfo);
                } else {
                    LOG.warning("No extension found for streamKey: %s (skipping)".formatted(streamKey));
                }
            }
        }
        PlatformServiceCache newCache = new PlatformServiceCache(
                createStreamKey(
                        platformCatalog.getRecommendedPlatform().getPlatformKey(),
                        platformCatalog.getRecommendedPlatform().getRecommendedStream().getId()),
                streams,
                platformCatalog,
                updatedStreamCatalogMap,
                LocalDateTime.now(ZoneOffset.UTC),
                platformTimestamp);

        checkNewCache(newCache);

        platformServiceCacheRef.set(newCache);
        Log.infof("""
                PlatformService cache has been reloaded successfully:
                platform timestamp: %s
                recommended stream key: %s (core: %s, platform: %s)
                recommended stream extensions: %d
                available streams: %s
                """.stripIndent(),
                platformTimestamp,
                recommendedStreamKey(),
                recommendedPlatformInfo().quarkusCoreVersion(),
                recommendedPlatformInfo().platformVersion(),
                recommendedCodeQuarkusExtensions().size(),
                String.join(", ", updatedStreamCatalogMap.keySet()));
    }

    private void checkNewCache(PlatformServiceCache newCache) throws IOException, QuarkusCommandException {
        // Only replace the existing values if we successfully fetched new values
        if (newCache.streamCatalogMap().isEmpty()) {
            throw new RuntimeException("No stream found");
        }

        // Check streams
        if (!newCache.streamCatalogMap().containsKey(newCache.recommendedStreamKey())) {
            throw new RuntimeException(
                    "Recommended stream not found in stream catalog: " + newCache.recommendedStreamKey());
        }

        if (LaunchMode.current().isDevOrTest()) {
            // skip the rest of the checks in dev/test mode
            return;
        }

        for (Map.Entry<String, PlatformInfo> entry : newCache.streamCatalogMap().entrySet()) {
            if (entry.getValue().codeQuarkusExtensions().isEmpty()) {
                throw new IllegalStateException("No extension found in the stream: " + entry.getKey());
            }
            projectService.createTmp(
                    entry.getValue(),
                    ProjectDefinition.builder().streamKey(entry.getKey())
                            .extensions(Set.of("resteasy", "resteasy-jackson", "hibernate-validator")).build(),
                    false,
                    true);

            Set<String> extensions = entry.getValue().extensionsById().containsKey("io.quarkus:quarkus-rest")
                    ? Set.of("rest", "rest-jackson", "hibernate-validator")
                    : Set.of("resteasy-reactive", "resteasy-reactive-jackson", "hibernate-validator");

            projectService.createTmp(
                    entry.getValue(),
                    ProjectDefinition.builder().streamKey(entry.getKey())
                            .extensions(extensions)
                            .build(),
                    false,
                    true);
            projectService.createTmp(
                    entry.getValue(),
                    ProjectDefinition.builder().streamKey(entry.getKey()).extensions(Set.of("spring-web")).build(),
                    false,
                    true);
        }
    }

    static String getStreamStatus(String version) {
        for (String part : version.split("[.-]")) {
            if (part.equalsIgnoreCase("redhat")) {
                return "FINAL";
            }
            if (!part.isBlank() && !isNumeric(part)) {
                return part.toUpperCase();
            }
        }
        return "FINAL";
    }

    private static boolean isNumeric(String part) {
        return NUMERIC_PATTERN.matcher(part).matches();
    }

    private String createStreamKey(String platformKey, String streamId) {
        return platformKey + SEPARATOR + streamId;
    }

    private String normalizeStreamKey(String streamKey) {
        if (streamKey == null) {
            return recommendedStreamKey();
        }
        return streamKey.contains(":") ? streamKey
                : createStreamKey(recommendedPlatformInfo().platformKey(), streamKey);
    }

    private static final Logger LOG = Logger.getLogger(PlatformService.class.getName());
    private static final String SEPARATOR = ":";

    public record PlatformServiceCache(
            String recommendedStreamKey,
            List<Stream> streams,
            PlatformCatalog platformCatalog,
            Map<String, PlatformInfo> streamCatalogMap,
            LocalDateTime cacheLastUpdated,
            String platformTimestamp) {
    }
}

File: api/src/main/kotlin/io/quarkus/code/service/PlatformService.kt
package io.quarkus.code.service

import io.quarkus.code.misc.QuarkusExtensionUtils.processExtensions
import io.quarkus.code.config.PlatformConfig
import io.quarkus.devtools.project.QuarkusProjectHelper
import io.quarkus.code.model.CodeQuarkusExtension
import io.quarkus.code.model.ProjectDefinition
import io.quarkus.code.model.Stream
import io.quarkus.devtools.project.JavaVersion
import io.quarkus.devtools.project.JavaVersion.getCompatibleLTSVersions
import io.quarkus.platform.catalog.processor.CatalogProcessor
import io.quarkus.platform.catalog.processor.CatalogProcessor.getMinimumJavaVersion
import io.quarkus.platform.catalog.processor.CatalogProcessor.getRecommendedJavaVersion
import io.quarkus.registry.Constants
import java.util.HashMap
import io.quarkus.registry.catalog.PlatformCatalog
import java.time.LocalDateTime
import jakarta.enterprise.event.Observes
import io.quarkus.runtime.StartupEvent
import io.quarkus.scheduler.Scheduled
import java.time.ZoneOffset
import java.time.ZoneId
import io.quarkus.registry.RegistryResolutionException
import org.apache.maven.artifact.versioning.DefaultArtifactVersion
import java.util.concurrent.atomic.AtomicReference
import java.util.logging.Level
import kotlin.Throws
import java.util.logging.Logger
import jakarta.inject.Inject
import jakarta.inject.Singleton

@Singleton
class PlatformService {


    @Inject
    lateinit var platformConfig: PlatformConfig

    @Inject
    lateinit var projectService: QuarkusProjectService

    private val catalogResolver = QuarkusProjectHelper.getCatalogResolver()
    private val platformServiceCacheRef: AtomicReference<PlatformServiceCache> = AtomicReference()

    fun onStart(@Observes e: StartupEvent?) {
        reload()
    }

    @Scheduled(cron = "{io.quarkus.code.quarkus-platforms.reload-cron-expr}")
    fun reload() {
        try {
            reloadPlatformServiceCache()
        } catch (e: RegistryResolutionException) {
            LOG.log(Level.SEVERE, "Could not resolve catalogs [" + e.message + "]", e)
        } catch (e: Exception) {
            LOG.log(Level.SEVERE, "Could not reload catalogs [" + e.message + "]", e)
        }
    }

    val isLoaded: Boolean
        get() = platformServiceCacheRef.get() != null && !recommendedCodeQuarkusExtensions.isNullOrEmpty()

    val platformsCache: PlatformServiceCache
        get() = platformServiceCacheRef.get() ?: error("Platforms cache must not be used if not loaded")

    val recommendedPlatformInfo: PlatformInfo
        get() = getPlatformInfo()

    val cacheLastUpdated: LocalDateTime
        get() = platformsCache.cacheLastUpdated

    val platformCatalog: PlatformCatalog
        get() = platformsCache.platformCatalog

    val recommendedCodeQuarkusExtensions: List<CodeQuarkusExtension>
        get() = getCodeQuarkusExtensions(recommendedStreamKey)

    val recommendedStreamKey: String
        get() = platformsCache.recommendedStreamKey

    val streams: List<Stream>
        get() = platformServiceCacheRef.get().streams

    val streamKeys: Set<String>
        get() = platformServiceCacheRef.get().streamCatalogMap.keys

    fun getCodeQuarkusExtensions(platformKey: String, streamId: String): List<CodeQuarkusExtension>? {
        val key = createStreamKey(platformKey, streamId)
        return getCodeQuarkusExtensions(key)
    }

    fun getCodeQuarkusExtensions(streamKey: String): List<CodeQuarkusExtension> {
        return getPlatformInfo(streamKey).codeQuarkusExtensions
    }

    fun getPlatformInfo(platformKey: String, streamId: String): PlatformInfo? {
        val key = createStreamKey(platformKey, streamId)
        return getPlatformInfo(key)
    }

    fun getPlatformInfo(streamKey: String? = null): PlatformInfo {
        val normalizedStreamKey = normalizeStreamKey(streamKey)
        return if (platformServiceCacheRef.get().streamCatalogMap.containsKey(normalizedStreamKey)) {
            platformServiceCacheRef.get().streamCatalogMap[normalizedStreamKey]!!
        } else throw IllegalArgumentException("Invalid streamKey: $streamKey")
    }

    @Throws(RegistryResolutionException::class)
    private fun reloadPlatformServiceCache() {
        catalogResolver.clearRegistryCache()
        val platformCatalog: PlatformCatalog? = if(platformConfig.registryId.isEmpty) catalogResolver.resolvePlatformCatalog()
            else catalogResolver.resolvePlatformCatalogFromRegistry(platformConfig.registryId.get())
        val updatedStreamCatalogMap: MutableMap<String, PlatformInfo> = HashMap()
        if(platformCatalog?.metadata == null || platformCatalog.platforms == null) {
            throw error("Platform catalog not found");
        }

        val platformTimestamp: String? = platformCatalog.metadata[Constants.LAST_UPDATED]?.toString()
        if(platformTimestamp.isNullOrBlank()) {
            throw error("Platform last updated date is empty");
        }
        if (platformServiceCacheRef.get()?.platformTimestamp == platformTimestamp) {
           LOG.log(Level.INFO, "The platform cache is up to date with the registry")
           return
        }
        val platforms = platformCatalog.platforms
        val streams = arrayListOf<Stream>()
        for (platform in platforms) {
            for (stream in platform.streams) {
                // Stream Map
                val recommendedRelease = stream.recommendedRelease
                val extensionCatalog = catalogResolver.resolveExtensionCatalog(recommendedRelease.memberBoms)
                val codeQuarkusExtensions: List<CodeQuarkusExtension> =
                    processExtensions(extensionCatalog)
                val platformKey = platform.platformKey
                val streamId = stream.id
                val streamKey = createStreamKey(platformKey, streamId)
                val lts = stream.metadata["lts"] as Boolean
                val minimumJavaVersion = getMinimumJavaVersion(extensionCatalog)

                val compatibleJavaLTSVersions = getCompatibleLTSVersions(JavaVersion(minimumJavaVersion))
                if (platformKey.equals("com.redhat.quarkus.platform")) {
                    // Hack to remove 21 support from code.quarkus.redhat.com
                    compatibleJavaLTSVersions.remove(21)
                }
                val recommendedJavaVersion = getRecommendedJavaVersion(extensionCatalog)?.toInt() ?: compatibleJavaLTSVersions.first()
                val quarkusCoreVersion = stream.recommendedRelease.quarkusCoreVersion
                val recommended = stream.id == platform.recommendedStream.id
                val streamInfo = Stream(
                    key = streamKey,
                    quarkusCoreVersion = quarkusCoreVersion,
                    platformVersion = stream.recommendedRelease.version.toString(),
                    recommended = recommended,
                    status = getStreamStatus(quarkusCoreVersion),
                    lts = lts,
                    javaCompatibility = Stream.JavaCompatibility(compatibleJavaLTSVersions, recommendedJavaVersion)
                )
                val platformInfo = PlatformInfo(
                    platformKey = platformKey,
                    quarkusCoreVersion = quarkusCoreVersion,
                    platformVersion = stream.recommendedRelease.version.toString(),
                    recommended = recommended,
                    codeQuarkusExtensions = codeQuarkusExtensions,
                    extensionCatalog = extensionCatalog,
                    stream = streamInfo
                )
                streams.add(streamInfo)
                updatedStreamCatalogMap[streamKey] = platformInfo
            }
        }
        val newCache = PlatformServiceCache(
            recommendedStreamKey = createStreamKey(
                platformCatalog.recommendedPlatform.platformKey,
                platformCatalog.recommendedPlatform.recommendedStream.id
            ),
            platformCatalog = platformCatalog,
            streamCatalogMap = updatedStreamCatalogMap,
            cacheLastUpdated = LocalDateTime.now(ZoneOffset.UTC as ZoneId),
            platformTimestamp = platformTimestamp,
            streams = streams
        )

        checkNewCache(newCache)

        platformServiceCacheRef.set(newCache)
        LOG.log(Level.INFO) {
            """
            PlatformService cache has been reloaded successfully:
                platform timestamp: $platformTimestamp
                recommended stream key: $recommendedStreamKey (core: ${recommendedPlatformInfo.quarkusCoreVersion}, platform: ${recommendedPlatformInfo.platformVersion})
                recommended stream extensions: ${recommendedCodeQuarkusExtensions.size}
                available streams: ${updatedStreamCatalogMap.keys.joinToString(", ")}
        """.trimIndent()
        }
    }

    private fun checkNewCache(newCache: PlatformServiceCache) {
        // Only replace the existing values if we successfully fetched new values
        if (newCache.streamCatalogMap.isEmpty()) {
            throw error("No stream found")
        }

        // Check streams
        if (!newCache.streamCatalogMap.containsKey(newCache.recommendedStreamKey)) {
            throw error("Recommended stream not found in stream catalog: " + newCache.recommendedStreamKey)
        }

        for (entry in newCache.streamCatalogMap) {
            if (entry.value.codeQuarkusExtensions.isNullOrEmpty()) {
                throw error("No extension found in the stream: " + entry.key)
            }
            projectService.createTmp(
                platformInfo = entry.value,
                projectDefinition = ProjectDefinition(
                    streamKey = entry.key,
                    extensions = hashSetOf("resteasy", "resteasy-jackson", "hibernate-validator")
                ),
                isGitHub = false,
                silent = true
            )
            projectService.createTmp(
                platformInfo = entry.value,
                projectDefinition = ProjectDefinition(
                    streamKey = entry.key,
                    extensions = hashSetOf("resteasy-reactive", "resteasy-reactive-jackson", "hibernate-validator")
                ),
                isGitHub = false,
                silent = true
            )
            projectService.createTmp(
                platformInfo = entry.value,
                projectDefinition = ProjectDefinition(streamKey = entry.key, extensions = hashSetOf("spring-web")),
                isGitHub = false,
                silent = true
            )
        }

    }

    private fun getStreamStatus(quarkusCoreVersion: String): String {
        val qualifier = DefaultArtifactVersion(quarkusCoreVersion).qualifier?.uppercase()
        return if (qualifier.isNullOrBlank())  "FINAL" else qualifier
    }

    private fun createStreamKey(platformKey: String, streamId: String): String {
        return platformKey + SEPARATOR + streamId
    }

    private fun normalizeStreamKey(streamKey: String?): String {
        if (streamKey == null) {
            return recommendedStreamKey
        }
        return if (streamKey.contains(":")) streamKey else createStreamKey(
            recommendedPlatformInfo.platformKey,
            streamKey
        )
    }

    companion object {
        private val LOG = Logger.getLogger(PlatformService::class.java.name)
        const val SEPARATOR = ":"
    }

    data class PlatformServiceCache(
        val recommendedStreamKey: String,
        val streams: List<Stream>,
        val platformCatalog: PlatformCatalog,
        val streamCatalogMap: MutableMap<String, PlatformInfo>,
        val cacheLastUpdated: LocalDateTime,
        val platformTimestamp: String
    )
}